Gemify

lukeasrodgers opened the pull request:
Converts this repo into a gem.

Still working on integration with our rails app, making sure migration generation works, etc.

lukeasrodgers committed:
Initial work on gemification.

lukeasrodgers committed:
Adding rspec skeleton.
Rspec warnings are disabled due to a bunch of annoying warnings
generated by the rspec-client gem, not related to our code.

lukeasrodgers committed:
Adding specs.

lukeasrodgers committed:
Adding `Migration.index_exists?` method.

lukeasrodgers committed:
Switching argument order in migrator for consistency with other methods.
Also raise exception if migraiton file not found, since we can't
continue in this case.

lukeasrodgers committed:
Changes to migrator proxy class.
* No need for `proxy` instance method.
* Avoid `method_missing` for now, not really needed, but some
metaprogramming could be use to DRY this class up.

lukeasrodgers committed:
Significant rewrite of migration code, added specs.
This tool should now support arbitrary migrations (i.e. run all
migrations that have not been run yet), migration to a specific
version, and rollback. Versions are tracked in a mandatory
`schema_versions` class on the database.

lukeasrodgers committed:
adding rake tasks to migrate, rollback, generate migration

lukeasrodgers committed:
Use `climate_control` gem to run some tess with modified ENV vars.

lukeasrodgers committed:
Allow specific schema version to be specified by rake migration task.

lukeasrodgers committed:
Adding some spec helper methods to reduce boilerplate.

lukeasrodgers committed:
Allow integration specs to be run via tag, rather than folder location.

lukeasrodgers committed:
Increase json gem dependency version.
Otherwise we get a conflict with the version required in our rails app.

lukeasrodgers commented:
@madnoodlesmlin I don't think we are actually using the `json` gem anywhere. Can we remove it as a dependency? Or maybe just make it a development dependency, if necessary?

madnoodlesmlin commented:
@lukeasrodgers since orientdb4r gem uses json 1.7.5 and on my machine I have multiple versions of json gem and by default it uses version 1.8.1, I just specify it so there will not have any conflict.

lukeasrodgers committed:
Rails app cannot find rake task, attempting to fix.

lukeasrodgers committed:
Fixing migration rake task, adding spec.
We need to coerce `Rails.root` to a string, otherwise we get stray
backslashes in our file path.

lukeasrodgers commented:
@madnoodlesmlin I don't see anywhere that orientdb4r requires json 1.7.5. Actually, their use of it seems a bit odd: they have a commented out statement requiring 1.5.1, so I guess they're actually just using ruby's built-in json module.

lukeasrodgers committed:
wip on handling connect/disconnect

lukeasrodgers committed:
Ensure that migration tasks are connected to db as necessary.

madnoodlesmlin commented:
@lukeasrodgers uh I see.

lukeasrodgers committed:
Environment should not be hardcoded in rake task (obviously).

lukeasrodgers committed:
Make sure Orientdb connection is terminated even if we get an exception.

madnoodlesmlin commented on lib/orientdb_schema_migrator/migration.rb:
```diff
@@ -40,33 +40,31 @@ def self.rename_class old_name, new_name
 
     def self.alter_property class_name, property_name, attribute_name, new_value
+      assert_class_exists!(class_name)
+      return false unless property_exists?(class_name, property_name)
```
I think we might want to change all validation check to use methods
assert_class_exists!(class_name)
assert_property_exists!(class_name, property_name)
Is there any reason that we want this method to return false?

lukeasrodgers replied:
Ah no, you're right, I'll fix that.

lukeasrodgers committed:
Raise exception if property we attempt to alter/drop does not exist.

lukeasrodgers committed:
Removing unused json dependency.

madnoodlesmlin commented on lib/orientdb_schema_migrator/migration.rb:
```diff
@@ -40,33 +40,31 @@ def self.rename_class old_name, new_name
         return true
       else
-        puts "#{class_name} doesn't exist"
         return false
```
do we want to change line 5,22,and here to use assert_class_exists! method?

akachrislee commented on lib/orientdb_schema_migrator/migration.rb:
```diff
@@ -40,33 +40,31 @@ def self.rename_class old_name, new_name
-      end
+    def self.add_index class_name, property_name, index_name, type
+      assert_class_exists!(class_name)
+      assert_property_exists!(class_name, property_name)
```
does ODB complain if we're trying to add an index that already exists?
i.e. should we be checking `index_exists?` before creating the index?

akachrislee commented on lib/orientdb_schema_migrator/migration.rb:
```diff
@@ -40,33 +40,31 @@ def self.rename_class old_name, new_name
+      ODBClient.command "create index #{index_name} on #{class_name} (#{property_name}) #{type}"
     end
 
     def self.drop_index index_name
```
does ODB complain if we're trying to remove an index that DOESN'T exists?
i.e. should we be checking `index_exists?` before removing the index?

lukeasrodgers replied:
Good question(s) -- I think ODB doesn't care in these situations, but I will double check.

madnoodlesmlin replied:
I think we should add a check for index when we create index. We don't have
to check when we drop index.

On Tue, Apr 7, 2015 at 2:05 PM, Luke Rodgers <notifications@github.com>
wrote:

> In lib/orientdb_schema_migrator/migration.rb
> <https://github.com/Headlinerfm/orientdb-schema-migrator/pull/1#discussion_r27903879>
> :
>
> >        ODBClient.command "alter property #{class_name}.#{property_name} #{attribute_name} #{new_value}"
> >      end
> >
> > -    def self.add_index class_name, index_name, property_name, type
> > -      # schema has to exist.
> > -      # check if property exists first, if not ask to create property.
> > -      if property_exists? class_name, property_name
> > -        ODBClient.command "create index #{index_name} on #{class_name} (#{property_name}) #{type}"
> > -        return true
> > -      else
> > -        puts "please create property #{property_name} first"
> > -        return false
> > -      end
> > +    def self.add_index class_name, property_name, index_name, type
> > +      assert_class_exists!(class_name)
> > +      assert_property_exists!(class_name, property_name)
> > +      ODBClient.command "create index #{index_name} on #{class_name} (#{property_name}) #{type}"
> >      end
> >
> >      def self.drop_index index_name
>
> Good question(s) -- I think ODB doesn't care in these situations, but I
> will double check.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Headlinerfm/orientdb-schema-migrator/pull/1/files#r27903879>
> .
>

madnoodlesmlin commented on lib/orientdb_schema_migrator/migration.rb:
```diff
@@ -40,33 +40,31 @@ def self.rename_class old_name, new_name
 
-    def self.add_property class_name, property_name, type,property_options={}
+    def self.add_property class_name, property_name, type, property_options={}
+      assert_class_exists!(class_name)
```
@lukeasrodgers Do we want to check if property already exists?

madnoodlesmlin committed:
add check for class already exists when create class
add check for property already exists when add property
add check for index already exists when add index

madnoodlesmlin committed:
clean tmp folder after rspec runs

madnoodlesmlin committed:
add tmp folder for rspec test

lukeasrodgers commented:
@madnoodlesmlin those changes look good to me.

lukeasrodgers committed:
Better gemspec spec.files directive.
Don't include Gemfile(.lock) or specs in built gem.

lukeasrodgers committed:
Add gemspec version requirements for orientdb4r and activesupport.
We are using odb 0.5.1. All versions of activesupport after 2.2.1
support camelize and constantize, which are the inflections we're
using, but climate_control requires >= 3.0, so let's run with that.

;;

lukeasrodgers committed:
Updating README with testing/contribution info.

lukeasrodgers committed:
Change gem name to use hyphens, not underscore.
Not a big diff, but more consistent with general gem-naming conventions.

lukeasrodgers committed:
Include rake task in gemspec files.

lukeasrodgers committed:
renaming main file
should have been done with 12a1a28, I think.

lukeasrodgers committed:
Add rake task descriptions.
Maybe these are required for integration with rails?

lukeasrodgers committed:
Create migrations folder if it doesn't exist.

lukeasrodgers committed:
Some usage instructions in README.

lukeasrodgers commented:
@madnoodlesmlin do you wanna review my latest changes? I think if they look okay, this is ready to merge. Probably not perfect/missing features, but good enough, I think.

madnoodlesmlin commented:
looks good. :+1:
